protocol/contracts/zaps/GenericZap.sol

```diff
@@ -0,0 +1,694 @@
+      }
+    }    
+  }
+
+  function _transferResidual(
+    address _recipient,
+    address _tokenA,
+    address _tokenB,
+    uint256 _amountA,
+    uint256 _amountB,
+    uint256 _amountAActual,
+    uint256 _amountBActual
+  ) internal {
+    if (_amountA > _amountAActual) {
+      _transferToken(IERC20(_tokenA), _recipient, _amountA - _amountAActual);
+    }
+
+    if (_amountB > _amountBActual) {
+      _transferToken(IERC20(_tokenB), _recipient, _amountB - _amountBActual);
+    }
+  }
+
+  function _swapTokens(
+    address _pair,
+    address _fromToken,
+    uint256 _fromAmount,
+    uint256 _amountOutMin
+  ) internal returns (uint256 amountA, uint256 amountB) {
+    IUniswapV2Pair pair = IUniswapV2Pair(_pair);
+    address token0 = pair.token0();
+    address token1 = pair.token1();
+
+    (uint256 res0, uint256 res1,) = pair.getReserves();
+    if (_fromToken == token0) {
+      uint256 amountToSwap = Swap.calculateSwapInAmount(res0, _fromAmount);
+      //if no reserve or a new pair is created
+      if (amountToSwap == 0) amountToSwap = _fromAmount / 2;
+
+      amountB = _swapErc20ToErc20(
+        _fromToken,
+        token1,
+        amountToSwap,
+        _amountOutMin
+      );
+      amountA = _fromAmount - amountToSwap;
+    } else {
+      uint256 amountToSwap = Swap.calculateSwapInAmount(res1, _fromAmount);
+      //if no reserve or a new pair is created
+      if (amountToSwap == 0) amountToSwap = _fromAmount / 2;
+
+      amountA = _swapErc20ToErc20(
+        _fromToken,
+        token0,
+        amountToSwap,
+        _amountOutMin
+      );
+      amountB = _fromAmount - amountToSwap;
+    }
+  }
+
+  function _fillQuotePool(
+    address _fromToken, 
+    uint256 _fromAmount,
+    address[] memory _coins,
+    address _swapTarget,
+    bytes memory _swapData
+  ) internal returns (uint256, uint256){
+    uint256 valueToSend;
+    if (_fromToken == address(0)) {
+      require(
+          _fromAmount > 0 && msg.value == _fromAmount,
+          "Invalid _amount: Input ETH mismatch"
+      );
+      valueToSend = _fromAmount;
+    } else {
+      SafeERC20.safeIncreaseAllowance(IERC20(_fromToken), _swapTarget, _fromAmount);
+    }
+    uint256 nCoins = _coins.length;
+    uint256[] memory balancesBefore = new uint256[](nCoins);
+    uint256 i = 0;
+    for (; i<nCoins;) {
+      balancesBefore[i] = IERC20(_coins[i]).balanceOf(address(this));
+      unchecked { i++; }
+    }
+
+    Executable.execute(_swapTarget, valueToSend, _swapData);
+
+    uint256 tokenBoughtIndex = nCoins;
+    uint256 bal;
+    // reuse vars
+    for (i=0; i<nCoins;) {
+      bal = IERC20(_coins[i]).balanceOf(address(this));
+      if (bal > balancesBefore[i]) {
+        tokenBoughtIndex = i;
+        break;
+      }
+      unchecked { i++; }
+    }
+    require(tokenBoughtIndex != nCoins, "Invalid swap");
+
+    return (tokenBoughtIndex, bal - balancesBefore[tokenBoughtIndex]);
+  }
+
+  // function _depositEth(
```

Comment:
Not sure if you were in the process of cleaning up?

Reply:
missed that, was coding in a zombie state last night